<a href="https://colab.research.google.com/github/lakmg2007/SCALAR_LEARNINGS/blob/main/Assignment_Day_12_Hypothesis_testing_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Question 1:** An experiment was performed to compare the effectiveness of Ammonium Chloride and urea on the grain yield (in quintal per hectare) and the results are given in the arrays below:

#Ammonium_chloride = [13.4, 10.9, 11.2, 11.8, 14, 15.3, 14.2, 12.6, 17, 16.2, 16.5, 15.7]
#Urea = [12, 11.7, 10.7, 11.2, 14.8, 14.4, 13.9, 13.7, 16.9, 16, 15.6, 16]
#Conduct an appropriate test to compare the same with a 95% confidence level and choose the appropriate option below.

#**To Identify which test to use:**

#There is no direction for the difference (like “Ammonium is better” or “Urea is worse”). Instead, the question is whether there’s a difference in effectiveness — that’s non-directional.

#why a two-tailed test?
#Because  H₀: μ₁ = μ₂  and  H₁: μ₁ ≠ μ₂

#This means its open to any difference — it could be that Ammonium Chloride yields more or less than Urea. A two-tailed test allows for both possibilities.



#**When would you use a one-tailed test instead?**
#If your hypothesis was directional — for example: > “Ammonium Chloride increases grain yield compared to Urea”

#Then your alternative hypothesis would be: > H₁: μ₁ > μ₂ (one-tailed, right)

#That’s more powerful when your assumption is strongly justified, but risky if the effect goes the opposite way — you'd miss it.

#**why ttest_ind is considered and not ztest**
#Population Standard Deviations Are Unknown
#The key reason: we don’t know the population standard deviations for grain yields with Ammonium Chloride and Urea. The z-test assumes these are known and fixed, which is rarely the case in real-world experiments.

#Since we only have sample standard deviations, the t-test is the correct choice.

#Sample Sizes Are Small (n = 12)
#With both groups having only 12 observations:

#Central limit theorem doesn’t fully kick in

#We must account for extra uncertainty in estimating the standard errors → which is exactly what the t-distribution does with its heavier tails

In [5]:
from statsmodels.stats import weightstats as stests
from scipy import stats
from scipy.stats import ttest_ind

Ammonium_chloride = [13.4, 10.9, 11.2, 11.8, 14, 15.3, 14.2, 12.6, 17, 16.2, 16.5, 15.7]
Urea = [12, 11.7, 10.7, 11.2, 14.8, 14.4, 13.9, 13.7, 16.9, 16, 15.6, 16]
t_stat, p_value = ttest_ind(Ammonium_chloride, Urea, equal_var=False)
print(f"t_stat:{t_stat:.4f}")
print(f"p_value:{p_value:.4f}")

t_stat:0.1846
p_value:0.8552


In [6]:
from scipy.stats import ttest_ind

ammonium_chloride = [13.4, 10.9, 11.2, 11.8, 14, 15.3, 14.2, 12.6, 17, 16.2, 16.5, 15.7]
urea = [12, 11.7, 10.7, 11.2, 14.8, 14.4, 13.9, 13.7, 16.9, 16, 15.6, 16]

t_stat, p_value = ttest_ind(ammonium_chloride, urea, equal_var=False)

print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")

T-statistic: 0.1846
P-value: 0.8552


#**Question 2:** When smokers smoke, nicotine is transformed into cotinine, which can be tested.

#The average cotinine level in a group of 50 smokers was 243.5 ng ml.

#Assuming that the standard deviation is known to be 229.5 ng ml.

#Test the assertion that the mean cotinine level of all smokers is equal to 300.0 ng ml, at 95% confidence.

#Null Hypothesis (H₀): μ = 300.0

#Alternative Hypothesis (H₁): μ ≠ 300.0 This is a two-tailed test because we’re checking for any difference, not just greater or lesser.

In [9]:
import math
from scipy.stats import norm
from scipy.stats import norm

sample_mean = 243.5
hypothesized_mean = 300.0
std_dev = 229.5
sample_size = 50
confidence_level = 0.95

# Step 1: Calculate Standard Error
standard_error = std_dev / math.sqrt(sample_size)
print(standard_error)

# Step 2: Compute z-statistic
z_stat = (sample_mean - hypothesized_mean) / standard_error

# Step 3: Determine critical z-value for 95% confidence (two-tailed)
z_critical = norm.ppf(1 - (1 - confidence_level) / 2)
# Step 4: Decision
if abs(z_stat) > z_critical:
    result = "Reject the null hypothesis"
else:
    result = "Fail to reject the null hypothesis"

# Output results
print(f"Z-statistic: {z_stat:.4f}")
print(f"Critical z-value: ±{z_critical:.4f}")
print(f"Conclusion: {result}")


# Two-tailed p-value
p_value = 2 * norm.sf(abs(z_stat))  # sf = survival function = 1 - cdf

print(f"P-value: {p_value:.4f}")


32.45620125646253
Z-statistic: -1.7408
Critical z-value: ±1.9600
Conclusion: Fail to reject the null hypothesis
P-value: 0.0817


#so the answer is : P-value: 0.0817 , the mean cotinine level of all smokers is equal to 300.0 ng/ml.

#**Question 3:** A French cafe has historically maintained that their average daily pastry production is at most 500.

#With the installation of a new machine, they assert that the average daily pastry production has increased. The average number of pastries produced per day over a 70-day period was found to be 530.

#Assume that the population standard deviation for the pastries produced per day is 125.

#Perform a z-test with the critical z-value = 1.64 at the alpha (significance level) = 0.05 to evaluate if there's sufficient evidence to support their claim of the new machine producing more than 500 pastries daily.

#Note: Round off the z-score to two decimal places.

#**How to determine i should conduct one-tailed z-test?**
##**Step 1: Look at the Alternative Hypothesis (H₁)**
##One-tailed test is appropriate when you're testing for a **specific direction of change** (either greater than or less than).

##Example: “The new machine increases pastry production” → H₁: μ > μ₀

##Example: “The new medication reduces blood pressure” → H₁: μ < μ₀

##Two-tailed test is appropriate when you're testing for any **difference, regardless of direction**.

##Example: “The new method changes the average processing time” → H₁: μ ≠ μ₀

##**Step 2: Is the Standard Deviation Known?**
##**If population standard deviation is known**, and sample size is reasonably large (n ≥ 30), you can use a **z-test**.

##If **σ is unknown**, you'd go for a **t-test** instead.

#**Rule of Thumb:**
#> If your research question uses words like **“increase”** or **“decrease”**, think **one-tailed**.
#> If it says **“change**, **impact, or effect”** without **specifying direction**, use **two-tailed**.





#**Null Hypothesis:**
#H₀: μ ≤ 500 (The average number of pastries produced per day is at most 500)

#But because we're using a one-tailed z-test with a claim of increase, we usually set the null in the form:

#**H₀:μ = 500**
#This is the specific value we test against.

#**Alternative Hypothesis:**
#H₁: μ > 500 (The new machine has increased average daily production)

#This framing sets us up for a right-tailed test, where we only consider values significantly greater than 500 as evidence against the null.

In [10]:
import math
from scipy.stats import norm

# Given data
sample_mean = 530
population_mean = 500
std_dev = 125
sample_size = 70
alpha = 0.05
z_critical = 1.64  # Given in the problem

# Step 1: Calculate Standard Error
standard_error = std_dev / math.sqrt(sample_size)

# Step 2: Compute z-statistic
z_stat = (sample_mean - population_mean) / standard_error

# Step 3: Compare z-statistic to critical z-value
if z_stat > z_critical:
    conclusion = "Reject the null hypothesis: The new machine increased production."
else:
    conclusion = "Fail to reject the null hypothesis: Not enough evidence of increased production."

# Step 4: Output results
print(f"Z-statistic: {z_stat:.2f}")
print(f"Critical z-value: {z_critical}")
print(f"Conclusion: {conclusion}")


Z-statistic: 2.01
Critical z-value: 1.64
Conclusion: Reject the null hypothesis: The new machine increased production.


# **Question 4:** The Chai Point stall at Bengaluru airport estimates that each person visiting the store drinks an average of 1.7 small cups of tea.

#Assume a population standard deviation of 0.5 small cups. A sample of 30 customers collected over a few days averaged 1.85 small cups of tea per person.

#Test the claim using an appropriate test at an alpha = 0.05 significance value, with a critical z-score value of ±1.96.

#Note: Round off the z-score to two decimal places.

# **Step 1:** Set up the hypotheses
# Null Hypothesis (H₀): μ = 1.7

# Alternative Hypothesis (H₁): μ ≠ 1.7

# This is a two-tailed test because we’re testing for any change — not just an increase or decrease.



In [11]:
import math
from scipy.stats import norm

# Given data
sample_mean = 1.85
population_mean = 1.7
std_dev = 0.5
n = 30
alpha = 0.05
z_critical = 1.96  # for 95% confidence (two-tailed)

# Step 1: Calculate Standard Error
standard_error = std_dev / math.sqrt(n)

# Step 2: Compute z-statistic
z_stat = (sample_mean - population_mean) / standard_error

# Step 3: Determine p-value (two-tailed)
p_value = 2 * norm.sf(abs(z_stat))

# Step 4: Conclusion based on critical z-value
if abs(z_stat) > z_critical:
    conclusion = "Reject the null hypothesis: The average cups per customer has changed."
else:
    conclusion = "Fail to reject the null hypothesis: No significant change in average cups per customer."

# Output
print(f"Z-statistic: {z_stat:.2f}")
print(f"P-value: {p_value:.4f}")
print(f"Conclusion: {conclusion}")


Z-statistic: 1.64
P-value: 0.1003
Conclusion: Fail to reject the null hypothesis: No significant change in average cups per customer.


# **Question 5:** A country has a population average height of 65 inches with a standard deviation of 2.5. A person feels people from his state are shorter. He takes the average of 20 people and sees that it is 64.5.

# At a 5% significance level (or 95% confidence level), can we conclude that people from his state are shorter, using the Z-test? What is the p-value?

# **Step 1: Define the hypotheses**
# Null Hypothesis (H₀): μ = 65 (no difference in average height)

# Alternative Hypothesis (H₁): μ < 65 (people from the state are shorter)

# This is a left-tailed test at α = 0.05.

# **When to use CDF,PDF and SF**

# **Use cdf when you're looking for Left-tailed test::**
# > “What is the probability that a value is less than or equal to X?”

# So, in the context of hypothesis testing:

# Left-tailed test: You use norm.cdf(z) → Because you're calculating the area to the left of the z-score

# Example:
# from scipy.stats import norm
# p_value = norm.cdf(-1.5)  # area to the left of z = -1.5

# **Use sf when you're looking for right-tailed test:**
# What if you're doing a right-tailed test?
# Use the **Survival Function**: norm.sf(z) → It’s just 1 - cdf(z), which gives area to the right.

# python
# p_value = norm.sf(1.5)  # area to the right of z = 1.5

# **For two-tailed tests?**
# You calculate the probability of getting as extreme or more in both directions:

# python
# z = 1.5
# p_value = 2 * norm.sf(abs(z))  # same as: 2 * (1 - norm.cdf(abs(z)))

In [12]:
import math
from scipy.stats import norm

# Given values
population_mean = 65
sample_mean = 64.5
std_dev = 2.5
sample_size = 20
alpha = 0.05

# Step 1: Calculate Standard Error
standard_error = std_dev / math.sqrt(sample_size)

# Step 2: Calculate z-statistic
z_stat = (sample_mean - population_mean) / standard_error

# Step 3: Calculate p-value for left-tailed test
p_value = norm.cdf(z_stat)  # since it's a 'shorter than' claim (left-tailed)

# Step 4: Determine conclusion
if p_value < alpha:
    conclusion = "Reject the null hypothesis: People from this state are significantly shorter."
else:
    conclusion = "Fail to reject the null hypothesis: Not enough evidence to conclude people are shorter."

# Output
print(f"Z-statistic: {z_stat:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Conclusion: {conclusion}")


Z-statistic: -0.8944
P-value: 0.1855
Conclusion: Fail to reject the null hypothesis: Not enough evidence to conclude people are shorter.


# **Question6:** Samples of Body fat percentages of few gym going men and women are recorded.

# men = [13.3, 6.0, 20.0, 8.0, 14.0, 19.0, 18.0, 25.0, 16.0, 24.0, 15.0, 1.0, 15.0]
# women = [22.0, 16.0, 21.7, 21.0, 30.0, 26.0, 12.0, 23.2, 28.0, 23.0]
# Perform an appropriate test to check if the mean body fat percentage of men and women is statistically different.

# Assume the significance level to be 5%.

# Choose the correct option

# To compare whether the mean body fat percentages of men and women are statistically different, we’ll use a two-sample t-test (Welch's t-test). This is appropriate because:

# We’re comparing means of two independent samples.

# Sample sizes and variances appear unequal.

# Population standard deviations are unknown.

# Step 1: State the Hypotheses
# Null Hypothesis (H₀): μ₁ = μ₂ (no difference in mean body fat %)

# Alternative Hypothesis (H₁): μ₁ ≠ μ₂ (means are different)

# This is a two-tailed test at α = 0.05.

In [14]:
from scipy.stats import ttest_ind

men = [13.3, 6.0, 20.0, 8.0, 14.0, 19.0, 18.0, 25.0, 16.0, 24.0, 15.0, 1.0, 15.0]
women = [22.0, 16.0, 21.7, 21.0, 30.0, 26.0, 12.0, 23.2, 28.0, 23.0]

t_stat, p_value = ttest_ind(men, women, equal_var=False)

print(f"T-statistic: {t_stat:.2f}")
print(f"P-value: {p_value:.2f}")


T-statistic: -2.90
P-value: 0.01


# At a 5% significance level (α = 0.05), the p-value (0.0133) is less than 0.05, so we reject the null hypothesis.

# **> Conclusion: There is a statistically significant difference in mean body fat percentage between gym-going men and women.**

# **Question 7:** Samples of IQ scores are collected from two competing schools, as follows:

# school_1 = [115, 111, 112, 101, 95, 98, 100, 90, 89, 108]
# school_2 = [107, 103, 91, 99, 104, 98, 117, 113, 92, 96, 108, 115, 116, 88]
# Perform an appropriate test with a 5% significance level to check if there is any statistically significant difference in the mean IQ's of these schools.

# Choose the correct option below :

# To determine whether there's a statistically significant difference in the mean IQ scores between the two schools, we’ll use Welch’s t-test — ideal when comparing two independent samples with potentially unequal variances and sample sizes.

# Step 1: State the Hypotheses
# Null Hypothesis (H₀): μ₁ = μ₂ (no difference in mean IQ)

# Alternative Hypothesis (H₁): μ₁ ≠ μ₂ (means are different)

# This is a two-tailed test at α = 0.05.

In [15]:
from scipy.stats import ttest_ind

school_1 = [115, 111, 112, 101, 95, 98, 100, 90, 89, 108]
school_2 = [107, 103, 91, 99, 104, 98, 117, 113, 92, 96, 108, 115, 116, 88]

t_stat, p_value = ttest_ind(school_1, school_2, equal_var=False)

print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.4f}")


T-statistic: -0.3726
P-value: 0.7134


# At a 5% significance level, the p-value (0.7167) is much greater than 0.05, so we fail to reject the null hypothesis.

# **> Conclusion: There is no statistically significant difference in the mean IQ scores between the two schools.**